File: smart-contract/crates/voting-crypto/src/commitment.rs
use sha2::{Digest, Sha256};

/// Vote commitment per ADR 0003: SHA-256 over UTF-8 proposal_id ‖ 32-byte salt ‖
/// 32-byte voter pubkey ‖ UTF-8 option_id (no separators).
pub fn vote_commitment(
    proposal_id: &str,
    salt: &[u8; 32],
    voter_pubkey: &[u8; 32],
    option_id: &str,
) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(proposal_id.as_bytes());
    hasher.update(salt);
    hasher.update(voter_pubkey);
    hasher.update(option_id.as_bytes());
    hasher.finalize().into()
}

File: tools/voting-shared/src/commitment.ts
import { createHash } from "crypto";

/** SHA-256 vote commitment (ADR 0003), matches `voting-crypto::vote_commitment`. */
export function voteCommitment(
  proposalId: string,
  salt: Uint8Array,
  voterPubkey: Uint8Array,
  optionId: string
): Uint8Array {
  if (salt.length !== 32 || voterPubkey.length !== 32) {
    throw new Error("salt and voter pubkey must be 32 bytes");
  }
  return createHash("sha256")
    .update(Buffer.from(proposalId, "utf8"))
    .update(salt)
    .update(voterPubkey)
    .update(Buffer.from(optionId, "utf8"))
    .digest();
}
